source/leveltabframewidget.cpp

```diff
@@ -0,0 +1,232 @@
+{
+    for (int y = 0; y < MICRO_HEIGHT / 2; y++) {
+        for (int x = 0; x < MICRO_WIDTH; x++) {
+            if (frame->getPixel(x, y).isTransparent()) {
+                return prepareMsgTransparent(msg, x, y);
+            }
+        }
+    }
+    return true;
+}
+
+static bool validLeftTrapezoid(const D1GfxFrame *frame, QString &msg)
+{
+    return validBottomLeftTriangle(frame, msg) && validTopHalfSquare(frame, msg);
+}
+
+static bool validRightTrapezoid(const D1GfxFrame *frame, QString &msg)
+{
+    return validBottomRightTriangle(frame, msg) && validTopHalfSquare(frame, msg);
+}
+
+void LevelTabFrameWidget::validate()
+{
+    int frameIdx = this->levelCelView->getCurrentFrameIndex();
+    D1GfxFrame *frame = this->gfx->getFrame(frameIdx);
+
+    QString error, warning, tmp;
+
+    if (frame->getWidth() != MICRO_WIDTH) {
+        error = "Invalid width. Must be 32px wide.";
+    } else if (frame->getHeight() != MICRO_HEIGHT) {
+        error = "Invalid height. Must be 32px wide.";
+    } else {
+        switch (frame->getFrameType()) {
+        case D1CEL_FRAME_TYPE::Square:
+            validSquare(frame, error);
+            break;
+        case D1CEL_FRAME_TYPE::TransparentSquare:
+            if (validSquare(frame, tmp)) {
+                warning = "Suggested type: 'Square'";
+                break;
+            }
+            if (validLeftTriangle(frame, tmp)) {
+                warning = "Suggested type: 'Left Triangle'";
+                break;
+            }
+            if (validRightTriangle(frame, tmp)) {
+                warning = "Suggested type: 'Right Triangle'";
+                break;
+            }
+            if (validLeftTrapezoid(frame, tmp)) {
+                warning = "Suggested type: 'Left Trapezoid'";
+                break;
+            }
+            if (validRightTrapezoid(frame, tmp)) {
+                warning = "Suggested type: 'Right Trapezoid'";
+                break;
+            }
+            break;
+        case D1CEL_FRAME_TYPE::LeftTriangle:
+            validLeftTriangle(frame, error);
+            break;
+        case D1CEL_FRAME_TYPE::RightTriangle:
+            validRightTriangle(frame, error);
+            break;
+        case D1CEL_FRAME_TYPE::LeftTrapezoid:
+            validLeftTrapezoid(frame, error);
+            break;
+        case D1CEL_FRAME_TYPE::RightTrapezoid:
+            validRightTrapezoid(frame, error);
+            break;
+        }
+    }
+
+    if (!error.isEmpty()) {
+        this->ui->frameTypeMsgLabel->setText(error);
+        this->ui->frameTypeMsgLabel->setStyleSheet("color: rgb(255, 0, 0);");
+    } else if (!warning.isEmpty()) {
+        this->ui->frameTypeMsgLabel->setText(warning);
+        this->ui->frameTypeMsgLabel->setStyleSheet("color: rgb(0, 255, 0);");
+    } else {
+        this->ui->frameTypeMsgLabel->setText("");
+    }
+}
+
+void LevelTabFrameWidget::on_frameTypeComboBox_activated(int index)
```

Comment:
im thinking we should handle this automatically rather then letting it be up to the user

Reply:
why would you restrict the user? The same frame can be encoded in multiple ways without causing a problem.

Reply:
ideally we could take a full tile or subtile as the input an automatically decompose and de-duplicate it

Reply:
that could be a separate feature, but that does not mean we should restrict the options of the user.

Reply:
an additional argument against auto-type-assignment: if we do that, the chance is high that we are going to have inefficient frames, where transparent square is used quite often only because of a few bad pixels.
If they had to assign them manually, they could have noticed that something is wrong with the asset.

Reply:
> The user wont really care how there data is encoded,

If the user does not care about it, then it wont open this tab at all.